chemprop/data/utils.py

```diff
@@ -333,98 +373,166 @@ def get_data(path: str,
         gt_targets, lt_targets = None, None
 
     # Load data
-    with open(path) as f:
+    extension = os.path.splitext(path)[1]
+    if extension == '.csv':
+        f = open(path)
```

Comment:
`with open(path) as f:` is the preferred pattern because it doesn't require a subsequent `f.close()`. Not presently the case here, but if you had mulitple exits to a function, you would have to be careful to close f in each of them, and only once per exit even with complicated logic. The `with` version of the opening makes it easier to maintain going forward.

Reply:
I understand that you changed it because the pkl reader is opening a file in a different way. And if we can't find another good way to do it, then that is a good enough reason to use `f=open(path)`. But I think worth the asking, is there a way to group these blocks so that you can use the `with` pattern?

Reply:
It looks like `pd.read_pickle` can accept the buffer f as input so I think that simplifies all this and allows the `with` block right?

Reply:
`pd.read_pickle` can take a binary-mode f object and `csv.DictReader` takes a text-mode f object. Trying to see if there's a way you could have them both use the same f. I think you could do something like

```
if extension in ['.pkl', '.pckl', '.pickle']:
    read_mode = 'rb'
else:  # extension = 'csv'
    read_mode = 'r'
with open(path, read_mode) as f:
...
```

Reply:
Is there a specific reason for adding support for pickle files? I think text-based formats are much more preferable in general.

Note that using `pd.read_pickle` also changes the general behavior here, such that the file will be loaded into memory in its entirely before being processed. This may be problematic for really large datasets.

Reply:
@mliu49 the main reason would be to be consistent with atom and bond features which use pickled pandas to store the extra dimension for atom/bond. We want for the outputs of atom/bond predictions to be useable as features. We should probably add support for csv with lists in the atom/bond features inputs. Would we want to only support csv? We could say that csv is preferred for large datasets. 

In regards to the incremental loading, we might progressively load targets, but we load features all at once. And features, especially atom/bond features, are usually larger than the targets. With so many supplemental feature types now added, we might be due to refactor this function.